src/vm_cpu.rs

```diff
@@ -1026,14 +987,14 @@ impl<T: GuestStagePagingMode> Drop for ActiveVmCpu<'_, '_, '_, T> {
             p.restore();
         }
 
-        // Unwrap safe because vcpu is locked and running, so vcpu must be at `vcpu_id`.
-        let once_entry = self
-            .container
-            .inner
-            .get(self.vcpu.vcpu_id as usize)
-            .unwrap();
-        let entry = once_entry.get().unwrap();
-        let mut status = entry.status.write();
+        // Avoid the lock order iversion by manually dropping the VmCpu::arch lock before taking
+        // the VmCpu::status lock (although it is safe from deadlock since status is checked before
+        // acquiring arch).
+        //
+        // Safety: self.arch is valid for dropping and we do not access it beyond this point.
+        unsafe { ManuallyDrop::drop(&mut self.arch) };
```

Comment:
I'll send you a patch to consider to avoid this unsafe. Because drop order is guaranteed to be the order listed in the struct, we can add a guard type to do the status set below and avoid manually drop.

It makes the code harder to read if you're not used to following drop chains though so feel free to ignore it.

Reply:
thanks, i prefer that to `ManuallyDrop` -- applied